src/diff/mod.rs

```diff
@@ -1,36 +1,269 @@
+                                    old_entry_opt = old_iter.next();
+                                    new_entry_opt = new_iter.next();
+                                },
+                                (SubDirectory(new_dir), File(old_file)) => {
+                                    diff.add_created_files(new_entry, parent_path)?;
+                                    diff.add_deleted_file(old_file, &RefCell::borrow(parent_path));
+                                    old_entry_opt = old_iter.next();
+                                    new_entry_opt = new_iter.next();
+                                },
+                                (SubDirectory(new_dir), SubDirectory(old_dir)) => {
+                                    parent_path.borrow_mut().push(new_dir.label.clone());
+                                    Diff::collect_diff(old_dir.deref(), new_dir.deref(), parent_path, diff)?;
+                                    parent_path.borrow_mut().pop();
+                                    old_entry_opt = old_iter.next();
+                                    new_entry_opt = new_iter.next();
+                                },
+                                // need to skip Repos
+                                (Repo, Repo) => {
+                                    old_entry_opt = old_iter.next();
+                                    new_entry_opt = new_iter.next();
+                                },
+                                (Repo, _) => {
+                                    old_entry_opt = old_iter.next();
+                                },
+                                (_, Repo) => {
+                                    new_entry_opt = new_iter.next();
+                                }
+                            }
+                        }
+                    }
+                },
+                (Some(old_entry), None) => {
+                    diff.add_deleted_files(old_entry, parent_path)?;
+                    old_entry_opt = old_iter.next();
+                },
+                (None, Some(new_entry)) => {
+                    diff.add_created_files(new_entry, parent_path)?;
+                    new_entry_opt = new_iter.next();
+                },
+                (None, None) => break
+            }
+        }
+        Ok(())
+    }
+
+    // if entry is a file, then return this file,
+    // or a list of files in the directory tree otherwise
+    fn collect_files_from_entry<F, T>(entry: &DirectoryContents, parent_path: &Rc<RefCell<Vec<Label>>>, mapper: F)
+        -> Result<Vec<T>, String>
+        where F: Fn(&File, &Vec<Label>) -> T + Copy {
+
+        match entry {
+            DirectoryContents::SubDirectory(dir) => Diff::collect_files(dir, parent_path, mapper),
+            DirectoryContents::File(file) => {
+                let mapped = mapper(file, &RefCell::borrow(parent_path));
+                Ok(vec![mapped])
+            },
+            _ => Err(String::from("Unexpected entry type"))
```

Comment:
I don't think it's unreasonable to do `DirectoryContents::Repo => vec![]` here

Reply:
It's not unreasonable, but why? When (and if) the Repo type is possibly updated to reflect embedded repositories (like a git subtree, for instance), returning an empty vec for it will become a programming error. Why not fail early and increase the changes to find this in tests?

Reply:
Actually, I believe that would be accounted for in this design :) It would be a `SubDirectory` where one of its entries is a `Repo` :)

Reply:
What about the name? For instance, Git subtree is connected to a directory, so it needs a label in your model.

Reply:
Do you mean the `.git` directory?

For all intents and purposes of browsing a VCS file system we won't want to reveal what's in `.git` since it's treated as a special directory. For example, you don't see those folders in GitHub or Nest.

Reply:
Ah, so Repo is for `.git`? Sorry, I was thinking it means something else entirely. In this case your initial comment makes total sense.

Reply:
Ah my bad! I should have clarified sooner :grin: We got there in the end!